src/adapter.rs

```diff
@@ -188,16 +201,46 @@ impl Adapter for DieselAdapter {
             .get()
             .map_err(|err| Box::new(Error::PoolError(err)) as Box<dyn StdError>)?;
 
-        let new_rule = self.save_policy_line(ptype, rule);
+        if let Some(new_rule) = self.save_policy_line(ptype, rule) {
+            return diesel::insert_into(casbin_rules)
+                .values(&new_rule)
+                .execute(&conn)
+                .map(|n| n == 1)
+                .map_err(|err| Box::new(Error::DieselError(err)) as Box<dyn StdError>);
+        }
 
-        diesel::insert_into(casbin_rules)
-            .values(&new_rule)
-            .execute(&conn)
-            .and_then(|n| if n == 1 { Ok(true) } else { Ok(false) })
-            .map_err(|err| Box::new(Error::DieselError(err)) as Box<dyn StdError>)
+        Ok(false)
     }
 
-    async fn remove_policy(&self, _sec: &str, pt: &str, rule: Vec<&str>) -> Result<bool> {
+    async fn add_policies(
+        &mut self,
+        _sec: &str,
+        ptype: &str,
+        rules: Vec<Vec<&str>>,
+    ) -> Result<bool> {
+        use schema::casbin_rules::dsl::casbin_rules;
+
+        let conn = self
+            .pool
+            .get()
+            .map_err(|err| Box::new(Error::PoolError(err)) as Box<dyn StdError>)?;
+
+        let new_rules = rules
+            .into_iter()
+            .filter_map(|x: Vec<&str>| self.save_policy_line(ptype, x))
+            .collect::<Vec<NewCasbinRule>>();
+
+        conn.transaction::<_, DieselError, _>(|| {
+            diesel::insert_into(casbin_rules)
```

Comment:
Do we need the transaction here? I think Diesel will generate one query.

Reply:
I was also thinking about this. But I'm not really sure about it. I'm not sure it's one query or multiple. If it's one query we can then remove it.

Reply:
I'm also not 100% sure in every case :see_no_evil: